src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs

```diff
@@ -166,32 +152,49 @@ private void MarkPropertiesAsSkipped(string currentModelKey, ModelMetadata metad
             }
         }
 
-        private bool ValidateProperties(
+        private IList<IModelValidator> GetValidators(IModelValidatorProvider provider, ModelMetadata metadata)
+        {
+            var validatorProviderContext = new ModelValidatorProviderContext(metadata);
+            provider.GetValidators(validatorProviderContext);
+            return validatorProviderContext.Validators;
+        }
+
+        private bool ValidateChildNodes(
             string currentModelKey,
             ModelExplorer modelExplorer,
             ValidationContext validationContext)
         {
             var isValid = true;
+            ExpandValidationNodeAndGetValidators(validationContext, modelExplorer);
 
-            foreach (var property in modelExplorer.Metadata.Properties)
+            IList<IModelValidator> validators = null;
+            if (modelExplorer.Metadata.IsCollectionType)
             {
-                var propertyExplorer = modelExplorer.GetExplorerForProperty(property.PropertyName);
-                var propertyMetadata = propertyExplorer.Metadata;
+                var enumerableModel = (IEnumerable)modelExplorer.Model;
+                var elementType = GetElementType(enumerableModel.GetType());
+                var elementMetadata = _modelMetadataProvider.GetMetadataForType(elementType);
+                validators = GetValidators(validationContext.ModelValidationContext.ValidatorProvider, elementMetadata);
```

Comment:
Its not to do with deign-time vs runtime metadata, the only reason why this exists is you do not want to call the factories to get validators for every element in the collection. That was an optimization existing in web api and i am just carrying it forward.